nixos/modules/virtualisation/xen-domU.nix

```diff
@@ -1,18 +1,90 @@
 # Common configuration for Xen DomU NixOS virtual machines.
 
-{ ... }:
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.virtualisation.xen;
+in
 
 {
-  boot.loader.grub.device = "nodev";
 
-  boot.initrd.kernelModules =
-    [ "xen-blkfront" "xen-tpmfront" "xen-kbdfront" "xen-fbfront"
-      "xen-netfront" "xen-pcifront" "xen-scsifront"
+  ## Interface ##
+
+  options.virtualisation.xen.guest = {
+    enable = lib.options.mkEnableOption "the Xen Guest Agent daemon, for easier XenStore access inside unprivileged domains";
+
+    package = lib.options.mkPackageOption pkgs "Xen Guest Agent" { default = "xen-guest-agent"; };
+
+    recommendedSettings = lib.options.mkOption {
+      default = true;
+      example = false;
+      type = lib.types.bool;
+      description = "Whether to enable the recommended settings for dedicated Xen PV/PVH guests.";
+    };
+  };
+
+  ## Implementation ##
+
+  config = lib.modules.mkIf cfg.guest.enable {
+    assertions = [
+      {
+        assertion = !cfg.enable;
+        message = "The Xen domU module cannot be enabled in a Xen dom0.";
+      }
+      {
+        assertion = !config.services.xe-guest-utilities.enable;
+        message = "The XenServer guest utilities cannot be enabled alongside the Xen Guest Agent.";
+      }
     ];
 
-  # Send syslog messages to the Xen console.
-  services.syslogd.tty = "hvc0";
+    environment.systemPackages = [ cfg.guest.package ];
+
+    systemd = {
+      packages = [ cfg.guest.package ];
+      services.xen-guest-agent.wantedBy = [ "multi-user.target" ];
+    };
+
+    boot = lib.modules.mkIf cfg.guest.recommendedSettings {
+      # PV/PVH guests are booted by the hypervisor, so we don't need a bootloader.
+      loader.grub.device = "nodev";
+      # PV/PVH guests don't need a kernel or initrd, as they're provided by Xen.
+      kernel.enable = false;
+      initrd.enable = false;
+      # Load guest-specific Xen kernel modules.
+      initrd.kernelModules = [
+        "xen-blkfront"
+        "xen-tpmfront"
+        "xen-kbdfront"
+        "xen-fbfront"
+        "xen-netfront"
+        "xen-pcifront"
+        "xen-scsifront"
+        "xen_hcd"
+      ];
+      kernelModules = [
+        "xen_wdt"
+        "xen_balloon"
+        "xen_fbfront"
+        "drm_xen_front"
+        "xen_kbdfront"
+        "xen_tpmfront"
+        "snd_xen_front"
+        "xenfs"
+        "9pnet_xen"
+      ];
+    };
```

Comment:
Yeah, they're useful for any Xen guest, I presume. I imagine whoever is maintaining the `xe-guest-utilities` package/module would include these in that module as well.

But to be totally fair, since the rust tools are supposed to replace `xe-guest-utilities`, I don't think we're losing much on not supporting them. I'm even considering just opening a PR to remove them, as they're unmaintained and not really part of vanilla Xen.

Reply:
fair enough